### PR TITLE
Python: Fix bad join in `ScopeEntryDefinition`

### PR DESCRIPTION
**Pull request by tausbn**
Before:

```
Tuple counts for Essa::ScopeEntryDefinition#class#24e22a14#f/1@45e0d8dh after 10.5s:
2133368   ~1%     {2} r1 = Essa::TEssaNodeDefinition#24e22a14#ffff_03#join_rhs AND NOT Essa::ImplicitSubModuleDefinition#class#24e22a14#f(Lhs.1 'this')
534478950 ~0%     {2} r2 = JOIN r1 WITH Definitions::SsaSourceVariable::getScopeEntryDefinition#dispred#f0820431#ff ON FIRST 1 OUTPUT Lhs.1 'this', Rhs.1
581249    ~4%     {1} r3 = JOIN r2 WITH Essa::EssaNodeDefinition::getDefiningNode#dispred#f0820431#ff ON FIRST 2 OUTPUT Lhs.0 'this'
                return r3
```

Let's see if pushing the `getDefiningNode` join further up improves the number of intermediary tuples. (Intuitively it should, since there should only be one defining node for any given `EssaNodeDefinition`.)

To do this, we unbind the `this.getSourceVariable()` part, which encourages the compiler to put this join later.

After:

```
Tuple counts for Essa::ScopeEntryDefinition#class#24e22a14#f/1@30758cv4 after 300ms:
2133569 ~1%     {2} r1 = SCAN Essa::TEssaNodeDefinition#24e22a14#ffff OUTPUT In.0, In.3 'this'
2133368 ~1%     {2} r2 = r1 AND NOT Essa::ImplicitSubModuleDefinition#class#24e22a14#f(Lhs.1 'this')
2133368 ~0%     {2} r3 = JOIN r2 WITH Definitions::SsaSourceVariable#class#486534ab#f ON FIRST 1 OUTPUT Lhs.1 'this', Lhs.0
2133368 ~0%     {3} r4 = JOIN r3 WITH Essa::EssaNodeDefinition::getDefiningNode#dispred#f0820431#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1, Lhs.0 'this'
581249  ~4%     {1} r5 = JOIN r4 WITH Definitions::SsaSourceVariable::getScopeEntryDefinition#dispred#f0820431#ff ON FIRST 2 OUTPUT Lhs.2 'this'
                return r5
```

Much better (and our intuition is confirmed -- joining with `getDefiningNode` did not increase the number of tuples).

### CODE REVIEW

File: python/ql/lib/semmle/python/essa/Essa.qll [hunk 1]
```diff
@@ -632,7 +632,9 @@ class DeletionDefinition extends EssaNodeDefinition {
  */
 class ScopeEntryDefinition extends EssaNodeDefinition {
   ScopeEntryDefinition() {
-    this.getDefiningNode() = this.getSourceVariable().getScopeEntryDefinition() and
+    this.getDefiningNode() =
+      pragma[only_bind_into](pragma[only_bind_out](this.getSourceVariable()))
+          .getScopeEntryDefinition() and
```
Comment (yoff) [inline python/ql/lib/semmle/python/essa/Essa.qll:637]: I believe this should be sufficient:
```suggestion
    this.getDefiningNode() =
      pragma[only_bind_out](this.getSourceVariable())
          .getScopeEntryDefinition() and
```

I tried writing it out with variables, getting
```codeql
this.getDefiningNode() = x and
this.getSourceVariable() = y and
z.getScopeEntryDefinition() = w and
x = w and
y = z and
not this instanceof ImplicitSubModuleDefinition
```
and I think that the only constraint we want to add is `pragma[only_bind_out](y) = z`, saying that `y` must be bound before we bind `z`, i.e. that `this.getSourceVariable()` must be computed before we compute `getScopeEntryDefinition`. Folding back up gives the above suggestion.

**Review by yoff — CHANGES_REQUESTED** [review]
I made a simplifying suggestion. If you already tried this and know that it does not work, let me know and I will approve. Otherwise, it might be nice to test if it is sufficient as I think it expresses the intuition more directly.

File: python/ql/lib/semmle/python/essa/Essa.qll (hunk 1 above)
Comment (tausbn) [reply python/ql/lib/semmle/python/essa/Essa.qll:637]: Here's what it does.

Before:
```
EVALUATE NONRECURSIVE RELATION:
  Essa::ScopeEntryDefinition#class#24e22a14#f(unique numbered_tuple this) :- 
    SENTINEL Definitions::SsaSourceVariable#class#486534ab#f
    SENTINEL project#Essa::EssaNodeDefinition::definedBy#dispred#f0820431#fff
    SENTINEL Definitions::SsaSourceVariable::getScopeEntryDefinition#dispred#f0820431#ff
    {2} r1 = SCAN Essa::TEssaNodeDefinition#24e22a14#ffff OUTPUT In.0, In.3 'this'
    {2} r2 = r1 AND NOT Essa::ImplicitSubModuleDefinition#class#24e22a14#f(Lhs.1 'this')
    {2} r3 = JOIN r2 WITH Definitions::SsaSourceVariable#class#486534ab#f ON FIRST 1 OUTPUT Lhs.1 'this', Lhs.0
    {3} r4 = JOIN r3 WITH project#Essa::EssaNodeDefinition::definedBy#dispred#f0820431#fff ON FIRST 1 OUTPUT Lhs.1, Rhs.1, Lhs.0 'this'
    {1} r5 = JOIN r4 WITH Definitions::SsaSourceVariable::getScopeEntryDefinition#dispred#f0820431#ff ON FIRST 2 OUTPUT Lhs.2 'this'
    return r5
```

After:
```
EVALUATE NONRECURSIVE RELATION:
  Essa::ScopeEntryDefinition#class#24e22a14#f(unique numbered_tuple this) :- 
    SENTINEL project#Essa::EssaNodeDefinition::definedBy#dispred#f0820431#fff
    SENTINEL Definitions::SsaSourceVariable::getScopeEntryDefinition#dispred#f0820431#ff
    {2} r1 = SCAN Essa::TEssaNodeDefinition#24e22a14#ffff OUTPUT In.0, In.3 'this'
    {2} r2 = r1 AND NOT Essa::ImplicitSubModuleDefinition#class#24e22a14#f(Lhs.1 'this')
    {2} r3 = SCAN r2 OUTPUT In.1 'this', In.0
    {3} r4 = JOIN r3 WITH project#Essa::EssaNodeDefinition::definedBy#dispred#f0820431#fff ON FIRST 1 OUTPUT Lhs.1, Rhs.1, Lhs.0 'this'
    {1} r5 = JOIN r4 WITH Definitions::SsaSourceVariable::getScopeEntryDefinition#dispred#f0820431#ff ON FIRST 2 OUTPUT Lhs.2 'this'
    return r5
```

And the tuple counts:
```
Tuple counts for Essa#24e22a14::ScopeEntryDefinition#f/1@916337pg after 244ms:
2133569 ~1%     {2} r1 = SCAN Essa#24e22a14::TEssaNodeDefinition#ffff OUTPUT In.0, In.3 'this'
2133368 ~1%     {2} r2 = r1 AND NOT Essa#24e22a14::ImplicitSubModuleDefinition#f(Lhs.1 'this')
2133368 ~0%     {2} r3 = SCAN r2 OUTPUT In.1 'this', In.0
2133368 ~0%     {3} r4 = JOIN r3 WITH project#Essa#24e22a14::EssaNodeDefinition::definedBy#2#dispred#fff ON FIRST 1 OUTPUT Lhs.1, Rhs.1, Lhs.0 'this'
581249  ~4%     {1} r5 = JOIN r4 WITH Definitions#486534ab::SsaSourceVariable::getScopeEntryDefinition#0#dispred#ff ON FIRST 2 OUTPUT Lhs.2 'this'
                return r5
```
:tada: 
I'll push a commit with this change (the shorter line means it needs to be reformatted anyway).

**Review by yoff — APPROVED** [review]
Lovely, thank you for indulging me :-)

**Pull request merged**